subt/docker/robotika/ros/robotika/src/drone_height.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/python
+
+from __future__ import print_function
+import rospy
+from nav_msgs.msg import *
+from std_msgs.msg import *
+from sensor_msgs.msg import *
+from geometry_msgs.msg import *
+
+import pdb
+    
+HEIGHT = 1.5
+MAX_ANGULAR = 0.7
+lastScanDown = 0.0
+lastScanUp = 0.0
+lastOdom = None
+validOdom = True
+
+def scan_down_callback(scan):
+    global lastScanDown, publisherTwist
+    lastScanDown = scan
+
+def scan_up_callback(scan):
+    global lastScanUp, publisherTwist
+    lastScanUp = scan
+
+def twist_callback(cmd_vel):
+    global HEIGHT, lastScanDown, lastScanUp
+    
+    
+    if lastScanDown.ranges[0] > HEIGHT:
+        cmd_vel.linear.z = -0.1
+    else:
+        cmd_vel.linear.z = 0.1
+    """
+    if lastScanDown.ranges[0] < lastScanUp.ranges[0]:
+        cmd_vel.linear.z = 0.1
+    else:
+        cmd_vel.linear.z = -0.1
+    """
+    #this is because drone can't handle too aggressive rotation
+    if cmd_vel.angular.z > MAX_ANGULAR: 
+        cmd_vel.angular.z = MAX_ANGULAR
+    elif cmd_vel.angular.z < -MAX_ANGULAR:
+        cmd_vel.angular.z = -MAX_ANGULAR
+
+    #if not validOdom:
```

Comment:
For future: Should we have a localization system, that is allowed to fail, and an odometry system, which never fails and is locally sort of correct? That would allow us to actively return to a last known location.

Reply:
My 2c: That is the approach I prefer. The position from the localization system can jump while the frame-to-frame relative position should respect the physics of the robot movement and can be used for example for speed control.